### ISSUE TITLE
--noImplicitAny behavior on 2.1.x

### ISSUE THREAD

**Issue by navossoc** [Microsoft/TypeScript#13596]
**TypeScript Version:**  >= 2.1.x

**Code**

```ts
// test 1 (the problem itself)
let x;              // no value assignment, so it's an any variable

x = 10;             // inferred as number
x = x + 5;          // valid operation
// but x remains as any, so...
x = 'rafael ' + x;  // perfect valid too
// but won't make any sense to me (see the explanation on discussion)
console.log(x);     // outputs: "rafael 15"
// I got an unexpected behavior here.
// The compiler didn't warn me about x being treated as a type any variable.

// test 2 (what I'm expecting from a typed variable)
let y = 10;         // inferred as number because of the value assignment

y = y + 5;          // valid operation
y = 'rafael ' + y;  // error: Type 'string' is not assignable to type 'number'.
// y is a number, so it can't be add to a string in this context (working as expected)
```

**Run:**
``tsc --noImplicitAny code.ts``

**Expected behavior:**
Raise error on declarations with an implied any type.

**Actual behavior:**
Don't raise any error and still let variables implicit declared as type *any*.

**Discussion:**
Reading the documentation, on [Compiler Options](https://www.typescriptlang.org/docs/handbook/compiler-options.html) it says:

> --noImplicitAny
> **Raise error on** expressions and **declarations with an implied any type**.

On the test 1 example x is still declared as _any_, so according to docs it should raise an error.

I have read the [What's new](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html) for 2.1.
There is a mention about this on _Improved any Inference_. It says:
> One great benefit of this is that you’ll see way fewer implicit any errors when running with --noImplicitAny.

I think this new "smart infer" like on the example below can be used as a default behavior as it **alone** is a good thing and don't affect anything else.

```ts
let x;

// You can still assign anything you want to 'x'.
x = () => 42;

// After that last assignment, TypeScript 2.1 knows that 'x' has type '() => number'.
let y = x();

// Thanks to that, it will now tell you that you can't add a number to a function!
console.log(x + y);
//          ~~~~~
// Error! Operator '+' cannot be applied to types '() => number' and 'number'.

// TypeScript still allows you to assign anything you want to 'x'.
x = "Hello world!";

// But now it also knows that 'x' is a 'string'!
x.toLowerCase();
```
[Example extracted from What's New - TypeScript 2.1](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html)

However, in my humble opinion it is breaking the behavior of the old compiler option.

So, if you guys are planning leave it **as it is right now**, it at least should be added as new compiler option and leave *--noImplicitAny* as it was before.

**Comment by RyanCavanaugh** [comment]
> However, in my humble opinion it is breaking the behavior of the old compiler option.

We generally don't consider producing fewer errors to be a breaking change. Is there a particular code pattern where you want an error, but didn't get one?

**Comment by WoLfulus** [comment]
I think what he means is that something that was considered an error before, stopped being an error now. I agree with @navossoc that this could be kinda problematic if you really want to be sure that a variable is what it's intended to be.

I imagine this is done using AST analysis to infer the type of the variable and generates, but "noImplicitAny" for me (and the docs) means that you can't really have "any" unless specifiec.

A variable that is a string on line 1, and a number on line 2, its of type "any", even though on line one it's infered to be number, and on line 2 a string.

**Comment by navossoc** [comment]
Hello @RyanCavanaugh ...

I'm not sure if I make myself clear here...

> We generally don't consider producing fewer errors to be a breaking change.
1. I like the new way the compiler can help catch this errors and even produce "less errors" in some cases, I can't argue against that.

>  Is there a particular code pattern where you want an error, but didn't get one?
The first test case is a perfect example of it.
I have a variable of type any there and I want a error there.

As I said before, if the option is called **noImplicitAny** and the behavior on the docs says:
> **Raise error on** expressions and **declarations with an implied any type**.

I want to be able to achieve that, but right now I just can't do it because the compiler option has another behavior.

I want to use that option to say to my myself: 
- Hey don't forget to type your variables matte. Be strict as possible!

@WoLfulus exactly, the variable still remains as type *any*, otherwise it would not be possible to assign a number then a string to it.

**Comment by RyanCavanaugh** [comment]
So what's key here is that you will still get an error if you use the variable in a way where we observe it as `any`. In that regard it's strictly "better" because it's only not-an-error in cases where the `any`ness wasn't used anyway. This removes the inconsistency that
```ts
let x = 10;
foo(x);
```
and
```ts
let x;
x = 10;
foo(x);
```
behaved differently.

If you want the "old" behavior, you can opt in to a TSLint rule that forces variable declarations to either have a type annotation or an initializer.

**Comment by navossoc** [comment]
Ok, I agree to disagree here.

As you said:

> So what's key here is that you will still get an error if you use the variable in a way where we observe it as any.

Then that documentation should be updated to reflect the new behavior.

To something like this:
> --noImplicitAny
Raise error on expressions and declarations **used** with an implied any type.

At least do you agree with me on that?

**Comment by mhegazy** [comment]
Really `--noImplicitAny` is to "Raise an error when the compiler gives a declaration a type `any` implicitly".

This by definition means that the compiler can get "smarter" in the future and give less implicit `any` types, and thus less errors.

**Comment by navossoc** [comment]
@mhegazy In my opinion, your sentence makes the purpose of the ``--noImplicitAny`` clearer.
I think this can be closed now...

**Comment by F1LT3R** [comment]
> If you want the "old" behavior, you can opt in to a TSLint rule that forces variable declarations to either have a type annotation or an initializer.

@RyanCavanaugh, what is that TSLint rule?

**Comment by navossoc** [comment]
@F1LT3R I think he is probably talking about this one:

```json
    "rules": {
        "no-unsafe-any": true
    },
```
https://palantir.github.io/tslint/rules/no-unsafe-any/

At least it is the only one I can think about.
